src/uct/ib/rdmacm/rdmacm_cm.c

```diff
@@ -165,26 +166,29 @@ static ucs_status_t uct_rdmacm_cm_id_to_dev_addr(struct rdma_cm_id *cm_id,
          * that the remote peer is reachable to the local one */
         roce_info.ver         = UCT_IB_DEVICE_ROCE_ANY;
         roce_info.addr_family = 0;
-    } else if (qp_attr.ah_attr.is_global) {
-        /* IB global address */
+        gid_p                 = &qp_attr.ah_attr.grh.dgid;
+    } else {
         address_pack_flags = UCT_IB_ADDRESS_PACK_FLAG_INTERFACE_ID |
                              UCT_IB_ADDRESS_PACK_FLAG_SUBNET_PREFIX;
-    } else {
-        /* IB local address - need just LID */
-        address_pack_flags = 0;
+        gid_p              = &cm_id->route.addr.addr.ibaddr.dgid;
+        if (qp_attr.ah_attr.is_global) {
+            ucs_assert(!memcmp(&cm_id->route.addr.addr.ibaddr.dgid,
+                               &qp_attr.ah_attr.grh.dgid, sizeof(*gid_p)));
+        } else {
+            /* IB local address requires GID for correct reachability check */
```

Comment:
this comment seems redundant, we shouldn't have accesses qp_attr.ah_attr.grh.dgid in the first place in case qp_attr.ah_attr.is_global is 0

Reply:
BTW, maybe we could move this code outside 'if'? it seems to be correct for both RoCE and IB case?

Reply:
> this comment seems redundant

ok, will remove

> we shouldn't have accesses qp_attr.ah_attr.grh.dgid in the first place in case qp_attr.ah_attr.is_global is 0

do you mean `LINK_LAYER_ETHERNET` case? there is a `ucs_assert(qp_attr.ah_attr.is_global);` 
so we don't

Reply:
> BTW, maybe we could move this code outside 'if'? it seems to be correct for both RoCE and IB case?

ok

Reply:
> do you mean `LINK_LAYER_ETHERNET` case? there is a `ucs_assert(qp_attr.ah_attr.is_global);`
> so we don't

I mean IB case: the essense of your fix is that we used qp_attr.ah_attr.grh.dgid in line 177 without checking it's valid.